streamz/core.py

```diff
@@ -1047,18 +1056,33 @@ class unique(Stream):
     3
     """
     def __init__(self, upstream, history=None, key=identity, **kwargs):
-        self.seen = dict()
+        self.seen = None
         self.key = key
-        if history:
-            from zict import LRU
-            self.seen = LRU(history, self.seen)
+        self.history = history
 
         Stream.__init__(self, upstream, **kwargs)
 
     def update(self, x, who=None):
         y = self.key(x)
+        # If this is the first piece of data make the cache
+        if self.seen is None:
+            if isinstance(y, Hashable):
+                self.seen = dict()
+                if self.history:
+                    # if it is hashable use LRU cache
+                    if isinstance(y, Hashable):
+                        from zict import LRU
+                        self.seen = LRU(self.history, self.seen)
+            # if not hashable use deque (since it doesn't need a hash)
+            else:
+                self.seen = deque(maxlen=self.history)
```

Comment:
With a deque, we get the specified number of items, not the specified number of *unique* items. I think that, if we are allowing now to process data to get something hashable, this branch is unnecessary.

Reply:
To be fair the ``key`` kwarg was always there, just not documented.

Reply:
The uniqueness is guaranteed by https://github.com/python-streamz/streamz/pull/244/files#diff-d4e512993f47710ea9e7155d66011a24R1080 not the `deque` itself.